packages/vscode-extension/src/debug/taskTerminal/baseTunnelTaskTerminal.ts

```diff
@@ -0,0 +1,273 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+/**
+ * @author Xiaofu Huang <xiaofhua@microsoft.com>
+ */
+import * as util from "util";
+import * as vscode from "vscode";
+
+import { assembleError, err, FxError, ok, Result, UserError } from "@microsoft/teamsfx-api";
+import { envUtil, isV3Enabled } from "@microsoft/teamsfx-core";
+import { LocalTelemetryReporter } from "@microsoft/teamsfx-core/build/common/local";
+import { pathUtils } from "@microsoft/teamsfx-core/build/component/utils/pathUtils";
+
+import VsCodeLogInstance from "../../commonlib/log";
+import { ExtensionErrors, ExtensionSource } from "../../error";
+import * as globalVariables from "../../globalVariables";
+import { ProgressHandler } from "../../progressHandler";
+import {
+  TelemetryEvent,
+  TelemetryProperty,
+  TelemetrySuccess,
+} from "../../telemetry/extTelemetryEvents";
+import { getDefaultString, localize } from "../../utils/localizeUtils";
+import { Step } from "../commonUtils";
+import { ngrokTunnelDisplayMessages, TunnelDisplayMessages } from "../constants";
+import { doctorConstant } from "../depsChecker/doctorConstant";
+import { localTelemetryReporter } from "../localTelemetryReporter";
+import { BaseTaskTerminal } from "./baseTaskTerminal";
+
+export interface IBaseTunnelArgs {
+  type?: string;
+  env?: string;
+  output?: {
+    endpoint?: string;
+    domain?: string;
+  };
+}
+
+export const TunnelType = Object.freeze({
+  devTunnel: "dev-tunnel",
+  ngrok: "ngrok",
+});
+
+export type OutputInfo = {
+  file: string | undefined;
+  keys: string[];
+};
+
+export type EndpointInfo = {
+  src: string;
+  dist: string;
+};
+
+export abstract class BaseTunnelTaskTerminal extends BaseTaskTerminal {
+  protected static tunnelTaskTerminals: Map<string, BaseTunnelTaskTerminal> = new Map<
+    string,
+    BaseTunnelTaskTerminal
+  >();
+
+  protected outputMessageList: string[];
+
+  protected readonly progressHandler: ProgressHandler;
+  protected readonly step: Step;
+
+  constructor(taskDefinition: vscode.TaskDefinition, stepNumber: number) {
+    super(taskDefinition);
+
+    for (const task of BaseTunnelTaskTerminal.tunnelTaskTerminals.values()) {
+      task.close();
```

Comment:
Should we delete the object in the map, otherwise its size will increase forever?

Reply:
When 'close' is called, the 'stop' will also be called, and the object will be removed in `stop`.

https://github.com/OfficeDev/TeamsFx/blob/3be41cef053d63178a57506a2ae4a42e52a39765/packages/vscode-extension/src/debug/taskTerminal/localTunnelTaskTerminal.ts#L117